### ISSUE TITLE
get nan on TPU but reasonable result on GPU

### ISSUE THREAD

**Issue by kkissmart** [pytorch/xla#2350]
## ❓ Questions and Help
I am running exact the same code with the same data and seed on TPU and GPU

on TPU, I got this
2020-07-15 19:00:31 | INFO | train_inner | epoch 001:      1 / 85212 loss=nan, ppl=nan, wps=0, ups=0, wpb=31472, bsz=32, num_updates=1, lr=1e-08, gnorm=nan, train_wall=6, wall=87

on GPU I got this 
2020-07-15 19:07:08 | INFO | train_inner | epoch 001:      1 / 85212 loss=16.796, ppl=113819, wps=0, ups=0, wpb=1904, bsz=32, num_updates=1, lr=1e-08, gnorm=3.949, loss_scale=128, train_wall=0, wall=58

any idea why?

Thanks so much in advance!

**Comment by taylanbil** [comment]
Which fairseq model/task is this?

**Comment by kkissmart** [comment]
@taylanbil 
Transformer model
https://github.com/kkissmart/tpu_fairseq
I added a lot of assertions and looks like the nan is from forward Multihead attn module 
https://github.com/pytorch/fairseq/blob/ffecb4e3496379edf5ecae1483df5b7e0886c264/fairseq/modules/multihead_attention.py#L212
something wrong after this line. 
I guess it is because of the softmax.

Let me know what you think
Thanks!

**Comment by taylanbil** [comment]
There are 14 lines in those asserts, could you pinpoint the line where the nan is introduced?

**Comment by kkissmart** [comment]
@taylanbil 
it failed here
https://github.com/kkissmart/tpu_fairseq/blob/f8ff9f22abdf33d6d13de525bf8aca08891102ff/fairseq/modules/multihead_attention.py#L360

**Comment by kkissmart** [comment]
actually https://github.com/kkissmart/tpu_fairseq/blob/f8ff9f22abdf33d6d13de525bf8aca08891102ff/fairseq/modules/multihead_attention.py#L358
norm is tensor(inf, device='xla:0', dtype=torch.bfloat16, grad_fn=<NormBackward0>)
so the softmax messed things up.

**Comment by taylanbil** [comment]
if it fails on 358, then it shouldn't be softmax. softmax output is set to `attn_weights_float`, not `attn_weights`. you're asserting on `attn_weights`. If you assert on line 354, it should also fail.

Does line 249 get executed ?https://github.com/kkissmart/tpu_fairseq/blob/f8ff9f22abdf33d6d13de525bf8aca08891102ff/fairseq/modules/multihead_attention.py#L349 It's inside an if statement so I cannot tell.

**Comment by kkissmart** [comment]
@taylanbil 
I had a typo on 358... my bad
after it fixed
it failed on 358
https://github.com/kkissmart/tpu_fairseq/blob/ea16d81231ce2bf17a75312e15dca505634923ce/fairseq/modules/multihead_attention.py#L359

but it made https://github.com/kkissmart/tpu_fairseq/blob/ea16d81231ce2bf17a75312e15dca505634923ce/fairseq/modules/multihead_attention.py#L355

any idea?

**Comment by kkissmart** [comment]
@taylanbil 
this is solved by add 
attn_weights = torch.clamp(attn_weights, min=-10000, max=10000)
before softmax
definitely not a good idea.
looks like a numerical stability issue.
Any idea on how to solve this?
Appreciate your help.

**Comment by taylanbil** [comment]
I'm looking into this. Trying to put together a smaller repro to see if there's a problem w/ softmax.

**Comment by taylanbil** [comment]
I tried for a while to see if softmax is problematic, but could not produce a nan in the output with very large inputs (outside the bounds of -10000 to 10000, but finite).

Then I put in infinity, -infinity in the input tensor. The result then has nan. But then I tried w/ regular pytorch, and I get the same behavior. So I'm not sure if there's anything to do on the torch_xla side here.

**Comment by kkissmart** [comment]
@taylanbil 
https://github.com/pytorch/fairseq/blob/77df83ab6e07468dbc8b31a5b84766f7fedd250d/fairseq/modules/multihead_attention.py#L340

multihead attention is designed to add float ('-inf') as mask. It will be problematic if it returns nan here. not sure how regular pytorch handled this.

**Comment by taylanbil** [comment]
Ah apologies, it only produces nan with input being `inf`. if input has `-inf`, output does not have nan. Behavior is consistent w/  regular pytorch, I verified.

So then I'm back to not being able to reproduce the issue. The links in [the comment above](https://github.com/pytorch/xla/issues/2350#issuecomment-659020386) are dead. You may have an inf in the input tensor maybe? can you assert not `torch.isinf(inputtensor).any()`

**Comment by kkissmart** [comment]
@taylanbil 

https://github.com/kkissmart/tpu_fairseq/blob/526b72951e1ed2c329424c78cdceccd5e7ff5f6d/fairseq/modules/multihead_attention.py#L343

no error msg on this line

**Comment by taylanbil** [comment]
at this point we'd need to understand which input is nan-ing. Could you run with one core `--distributed-world-size=1`, and insert a `pdb.set_trace()` when `attn_weights_float` has a `nan`, and then inspect the input that leads to it?

**Comment by kkissmart** [comment]
> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(
Exception in device=TPU:0: 
(Pdb) > /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(

Exception in device=TPU:5: 


(Pdb) (Pdb) (Pdb) 


> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(
(Pdb) Exception in device=TPU:4: 
> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(Exception in device=TPU:7: 


Exception in device=TPU:2: 
(Pdb) Exception in device=TPU:3:

**Comment by kkissmart** [comment]
> /home/yinhanliu/tpu_fairseq/fairseq/modules/multihead_attention.py(351)forward()
-> attn_weights_float = utils.softmax(

**Comment by kkissmart** [comment]
my bad... I had instances that have entire sentences filled with padding tokens.
When I add buckets to instances, I failed to catch instances that are empty before adding paddings.
